File: test/kmer_test1.c
#include "../src/kmers.h"
#include <string.h>

int main(int argc, char ** argv){

	char seq[13] = "ATCGGCTACTATT";

	fprintf(stderr, "Starting seq: %s\n", seq );

	char ans_lookups[6][13] = {
		"TCGGCTACTATT",
		"ATCGGCTACTAT",
		"TAGCCGATGATA",
		"AGCCGATGATAA"
	};

	pot_t * kms =  dna_to_kmer(seq, 13, 12);
	pot_t * qms =  dna_to_kmer(seq, 13, 12);

	fprintf(stderr, "%s\n", "kms" );
	print_pot(kms);
	fprintf(stderr, "\n\n" );
	fprintf(stderr, "%s\n", "qms" );
	print_pot(qms);

	make_searchable(kms);

	uint32_t i;

	int final_res = 0;

	for(i = 0; i < kms->n; i++) {
		char * tmp = bin_to_dna(kms->seeds[i].kmer, kms->word_size, kms->seeds[i].strand);
		int res = strcmp(tmp, ans_lookups[i]);

		fprintf(stderr, "[%s:DNA->BIT->DNA:TEST] kmer: %llu pos:%i expecting:%s observed:%s [%s]\n", __FILE__, kms->seeds[i].kmer, kms->seeds[i].pos, ans_lookups[i], tmp, res != 0 ? "FAIL" : "PASS" );

		if(res != 0 ) final_res = 1;

		free(tmp);
	}


	Stack * hits = search(kms, qms);

	seed_pair_t pair;

	while(!stackPop(hits, (void*)&pair)) {
		fprintf(stderr, "qb:%i tb:%i qs:%i ts:%i %s %s\n", pair.a->pos, pair.b->pos, pair.a->strand, pair.b->strand, bin_to_dna(pair.a->kmer, kms->word_size, pair.a->strand), bin_to_dna(pair.b->kmer, kms->word_size, pair.b->strand) );
	}

	return final_res;

}

File: test/kmer_test1.nim
# vim: sw=4 ts=4 sts=4 tw=0 et:
#from strformat import fmt
import deques
from os import nil
from strutils import format
from pbpkg/kmers import nil

proc main*(args: seq[string]): int =
    var sq = "ATCGGCTACTATT"

    echo format("Starting seq: $#", sq)

    var ans_lookups = [
        "AGCCGATGATAA",
        "TAGCCGATGATA",
        "ATCGGCTACTAT",
        "TCGGCTACTATT",
    ]
    var kms: kmers.pot_t = kmers.dna_to_kmers(sq, 12)
    var qms: kmers.pot_t = kmers.dna_to_kmers(sq, 12)

    echo "kms"
    kmers.print_pot(kms)
    echo ""
    echo "qms"
    kmers.print_pot(qms)

    discard kmers.make_searchable(kms)  # also sorts

    var final_res: int = 0
    var i: uint32 = 0

    while i < kms.n:
        let tmp = kmers.bin_to_dna(kms.seeds[i].kmer, kms.word_size,
                                   kms.seeds[i].strand)
        let res = cmp(tmp, ans_lookups[i])

        echo format("[$#:DNA->BIT->DNA:TEST] kmer: $# pos:$# expecting:$# observed:$# [$#]",
            system.currentSourcePath(), kms.seeds[i].kmer, kms.seeds[i].pos, ans_lookups[i], tmp,
                if res != 0: "FAIL" else: "PASS")

        final_res = final_res or res

        inc(i)

    var hits = kmers.search(kms, qms)

    try:
        while true:
            let pair = deques.popLast(hits)
            echo format("qb:$# tb:$# qs:$# ts:$# $# $#",
                pair.a.pos, pair.b.pos,
                pair.a.strand, pair.b.strand,
                kmers.bin_to_dna(pair.a.kmer, kms.word_size, pair.a.strand),
                kmers.bin_to_dna(pair.b.kmer, kms.word_size, pair.b.strand))
    except:
        discard

    return final_res

when isMainModule:
    quit main(os.commandLineParams())
